### PR TITLE
pdfreader and bamboohr paycheck importer

### LINKED ISSUE #93: pdfreader as a new file format reader
First, thank you so much for your blog, and your importers and everything. I've been using beancount for 3 years now, and its just gotten too much for me to input stuff. Your blog inspired me to try again and to be a lot smarter about it this time.

I had an idea for a `pdfreader` module. My thought was to utilize the pdfplumber package, to extract tables from pdfs, and something like [this filter](https://github.com/jsvine/pdfplumber/issues/242#issuecomment-1686505282) to capture all the data outside of the tables as well. I'd like to also search for everything that looks like a date in outside-of-table information, and return those in a list. And then everything that looks like a currency outside of the tables as well in a list. Perhaps we can put all that outside-of-table information into a metadata table and pass all of the data on as a multi table format to downstream formatters.

My primary concern right now is reading my paystub pdf, but I think it could also be useful for parsing pdf statements. Thoughts?

### PR DESCRIPTION
**Pull request by a0js**
## Purpose
Many paycheck downloads only come in the form of pdf. To support this use case, and potentially other use cases, we need to add support for pdf parsing. As a proof of concept, I also added the bambooHR paystub importer.

## Approach
- I mirrored the csv_multiple_reader.py as best I could and ended up primarily writing out the read_file method.
- I added a debug option to print out some helpful hints when using the pdfreader.
- For the BambooHR, I mostly copied the workday importer and added the few extra settings needed for pdf parsing.

## Outstanding Issues
- Tests
  - I can't really upload my own paystub as a test. I've tried asking BambooHR for a demo paystub pdf, but they just redirected me to [their website that has a png version of the file](https://help.bamboohr.com/s/article/1185885#what-does-a-downloaded-paystub-look-like). I actually have a friend that works at BambooHR, so I'm reaching out to see if he can send me a demo pdf that I could upload. That said, I do have it working on my machine, (I know that really isn't good enough, but it is something). I could also write another importer for adp, and may be able to snag a sample pdf for them, maybe.
- Dependencies
  - I mostly developed this in my own beancount setup, so I'm not sure how to add the necessary dependencies (pandas, pdfplumber, & dateparser). Could you help me figure out how to add those dependencies to the repo, if this is something you want to move forward with?

Resolves #93

### CODE REVIEW

**Comment by redstreet** [comment]
Looks great, and I'd be happy to take this in. I completely understand the challenge in adding any kind of examples at all to importers since by definition they're all personal data. But I'm wondering if perhaps adding an example pdf that you even made up in any word processing program that is importable by this? I just want to ensure we have some way down the line to test code to ensure it's still alive.

It doesn't have to use the bamboohr importer (though that'd be great too). Perhaps a simple single table made in MS-Word or the likes, and a simple dummy importer for it?

**Comment by redstreet** [comment]
For dependencies, the following should work:
```
pip3 install pigar
pigar generate -c '>='
```

**Comment by ranebrown** [comment]
I think the dependencies would just need added [here](https://github.com/redstreet/beancount_reds_importers/blob/main/setup.py#L28)

**Comment by a0js** [comment]
Apologies for not getting back to this. I'm hoping to make the changes and add a test before the end of the week.

File: beancount_reds_importers/libreader/pdfreader.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                        'header_map':self.header_map,
+                        'paycheck_template':self.paycheck_template
+                    }
+                }))
+
+         
+
+        self.alltables = {}
+        for table in tables:
+            self.alltables[table['section']] = etl.fromdataframe(pd.DataFrame(table['table'][1:], columns=table['table'][0]))
```
[the remark this answers is not in the document]
Comment (a0js) [reply beancount_reds_importers/libreader/pdfreader.py:182]: If we can have it work without `pandas` that would be great. I'll implement you suggestion and try to get a test case up before the end of the week.

File: requirements.txt
Comment (a0js) [inline requirements.txt:1]: I'm not sure why pigar changed the file so much. It appears to still work, though.

File: beancount_reds_importers/libreader/pdfreader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (a0js) [reply beancount_reds_importers/libreader/pdfreader.py]: Turns out we can just run `etl.wrap(table['table'])` rather than preprocessing it.

**Comment by a0js** [comment]
I added a generic pdf paycheck importer that really isn't meant to be used but to show how to make a paycheck importer using the pdfreader. Tests are included and passing.

File: beancount_reds_importers/importers/genericpdf/tests/genericpdf_test.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from os import path
+from beancount.ingest import regression_pytest as regtest
+from beancount_reds_importers.importers import genericpdf
+
+
+@regtest.with_importer(
+    genericpdf.Importer(
+        {
+            "desc": "Paycheck",
+            "main_account": "Income:Salary:FakeCompany",
```
Comment (a0js) [inline beancount_reds_importers/importers/genericpdf/tests/genericpdf_test.py:10]: Is this correct? Should the `main_account` be the `Income:Salary` account or the `Assets:Bank` account? I know this doesn't matter much but I'm not sure on the philosophy.

File: requirements.txt
Comment (redstreet) [reply requirements.txt:1]: It's moved all versions forward based on what is installed on your system. That's fine, I wouldn't worry about it much. `requirements.txt` is used only to setup development environments, including by the github action to build the beancount_reds_importers package.

The package's dependencies are in `setup.py` which you will have to update manually. See `install_requires` under it.

File: beancount_reds_importers/importers/genericpdf/tests/genericpdf_test.py (hunk 2 above)
Comment (redstreet) [reply beancount_reds_importers/importers/genericpdf/tests/genericpdf_test.py:10]: Yes, I'd prefer to leave it as it is to be consistent, but you're right in that it really doesn't matter.

The only thing this really affects is how you store your Beancount files: if you prefer to use a separate `Income.Salary.FakeCompany.beancount` file as a journal for all your paystub vs. if you prefer to stick these transactions into your `Assets.Bank.beancount` journal. Some users use one large journal file, in which case this makes no difference at all.

**Comment by redstreet** [comment]
This looks great thank you, and thanks for the test! Looks like the package build fails. If you could please take a look and fix that, we can get this merged in.

**Comment by a0js** [comment]
> I think the dependencies would just need added [here](https://github.com/redstreet/beancount_reds_importers/blob/main/setup.py#L28)

I forgot to do this. I'll update it tomorrow!

File: requirements.txt
Comment (a0js) [reply requirements.txt:1]: I think the build is failing because pigar removed some dependencies. I'll look into it tomorrow and see what I can do.

**Comment by a0js** [comment]
I wasn't able to get to this before I went off grid for a week. I'll try and update in the next few days.

**Comment by a0js** [comment]
@redstreet Not sure why, but I could not get pigar to find all the dependencies necessary, so I just manually updated the `requirements.txt` file. That should fix the tests.

**Comment by redstreet** [comment]
@a0js, curious, is there a button to run the formatting workflow that appears for you in this PR? I made a commit yesterday to enable this and am wondering if it works.

**Comment by redstreet** [comment]
> Not sure why, but when I ran the formatting commands locally I got 64 changes instead of 3 as observed in the github action output. I'm working on a mac and for some reason the default ruff settings are different. I ended up running a docker image that copied the github action environment and ran the format commands inside docker and that got the correct formatting changes.

Thanks for reporting. Hmm, not sure why you're seeing this. Ruff [uses](https://docs.astral.sh/ruff/configuration/) settings frm `pyproject.toml` that's in the repo. So if you're running from the repo root and that file is present, it should come from there. Perhaps try a `ruff --version`, and a `pip install ruff --ugprade` if needed?

> Thanks for your patience, by the way.

Of course, no worries at all, thank you for sticking with this PR and getting it in, much appreciated!

**Comment by a0js** [comment]
> @a0js, curious, is there a button to run the formatting workflow that appears for you in this PR? I made a commit yesterday to enable this and am wondering if it works.

I don't see anything on the PR page, but I could be looking in the wrong place. Where should it be?

**Comment by redstreet** [comment]
> I don't see anything on the PR page, but I could be looking in the wrong place. Where should it be?

I'd expect it to be on the bottom of the PR page. I'm sure it'd be an obvious green button, so if you haven't seen it, it must've not worked. Anyway, it doesn't matter much for this PR, thanks for checking!

Do let me know if you need help with any of the outstanding things.

**Comment by a0js** [comment]
I can't merge this myself as I don't have write access. If you think this is good to go, can you merge it in?

**Review by redstreet — APPROVED** [review]

**Comment by redstreet** [comment]
I think it was failing checks, but the checks are not running now for some reason. Let me take a look.

**Comment by redstreet** [comment]
Checks are passing for me locally. I don't know why github wouldn't run them. Either way, merged!

Thank you again for the contribution, and for working to get this PR through! IMO, table extraction from pdfs is a solid contribution for `beancount_reds_importers` as it's still fairly common to find that pdfs are the only option (no csv/ofx). So this is great!

**Comment by a0js** [comment]
You're most welcome! I'm glad I could add something to this awesome project. Let me know if there are some other features I can help with.

**Comment by a0js** [comment]
Sorry to comment on the old PR, but I was just curious how often you cut releases and when this one might be published?

**Comment by redstreet** [comment]
Np at all. I usually put it through personal use of at least a few weeks before I publish, so bugs have a chance to surface. Let me take a look this evening to see if I can cut a release.

**Comment by redstreet** [comment]
[Released 0.9.0](https://github.com/redstreet/beancount_reds_importers/releases/tag/0.9.0), featuring this PR :-)

**Pull request merged**